### PR TITLE
Using traditional MTX ingest for AnnData processed expression (SCP-5124)

### PR DESCRIPTION
**Pull request by bistline**
#### BACKGROUND & CHANGES
This update switches AnnData processed expression ingest to use extracted MTX files, instead of direct MongoDB inserts.  This has multiple advantages over the previous model, as it is faster and reuses existing/tested code.  A user can now upload a single AnnData file and extract all 3 primary data types: clustering, metadata, and processed expression.  Also, a user no longer receives the initial success email for the extraction job as it does not contain any actionable information, and could lead to confusion.  Now, users only receive emails for downstream, traditional parse jobs of extracted flat files.

In addition, this contains refactoring of all the ingest-related "parameter" classes that leverage the `Parameterizable` module, as well as some other method consolidations.  This was done to standardize interfaces across similar classes and to make adding new attributes easier in the future.

Note: there is code in place to launch subsampling jobs for `AnnData` files, but this work is blocked by SCP-5140.  The logic works as intended, but the resulting MongoDB index violations cause the jobs to fail and all data to be deleted.  Once these violations are dealt with, subsampling for `AnnData` files will operate the same as normal SCP-formatted flat files.

#### MANUAL TESTING
1. Boot as normal and start all services related to file parsing
2. Sign in and create a new study
3. Select the `AnnData` UX, and upload the file at `test/test_data/anndata_test.h5ad`
   * There is clustering available for `obsm_key_name: X_tsne`
4. After uploading the file, note in the logs that the initial `extract` param contains `processed_expression` along with `cluster` and `metadata`
5. Once the first job finishes, note that the parsing job for expression launches with the `mtx` parameters:
```
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 6451311b94ec8f59c754a6d7
  - "--study-file-id"
  - 646644d9b636f8cb7d4a6389
  - "--user-metrics-uuid"
  - 0fc4d8e5-6c4c-422e-94a3-c78801757202
  - ingest_expression
  - "--matrix-file"
  - gs://fc-38dbd25a-eea7-4273-849b-732b1e664249/_scp_internal/anndata_ingest/SCP91_646644d9b636f8cb7d4a6389/h5ad_frag.matrix.processed.mtx.gz
  - "--matrix-file-type"
  - mtx
  - "--gene-file"
  - gs://fc-38dbd25a-eea7-4273-849b-732b1e664249/_scp_internal/anndata_ingest/SCP91_646644d9b636f8cb7d4a6389/h5ad_frag.features.processed.tsv.gz
  - "--barcode-file"
  - gs://fc-38dbd25a-eea7-4273-849b-732b1e664249/_scp_internal/anndata_ingest/SCP91_646644d9b636f8cb7d4a6389/h5ad_frag.barcodes.processed.tsv.gz
  :flags: []
```
6. (Optional) Wait until all 3 jobs complete, and then load the study and confirm that visualizations are enabled, and you can query for genes like `Foo` and `Bar`:
Screenshot 2023-05-18 at 5 00 30 PM

### CODE REVIEW

File: app/models/ingest_job.rb [hunk 1]
```diff
@@ -633,38 +664,49 @@ def set_anndata_file_info
...
     # reference AnnData uploads don't have extract parameter so exit immediately
-    return nil if params_object.extract.blank?
+    return if params_object.extract.blank?
 
     params_object.extract.each do |extract|
       case extract
       when 'cluster'
         params_object.obsm_keys.each do |fragment|
-          Rails.logger.info "launching AnnData #{fragment} cluster extraction for #{study_file.upload_file_name}"
+          Rails.logger.info "launching AnnData #{fragment} cluster ingest for #{study_file.upload_file_name}"
```
Comment (eweitz) [inline app/models/ingest_job.rb:673]:
```suggestion
          Rails.logger.info "Launching AnnData #{fragment} cluster ingest for #{study_file.upload_file_name}"
```

I see some Rails log messages in this file start with a capital letter, and some don't.  For easy naive greppability, which helps debuggability, it'd be ideal to consistently case the first letter here and elsewhere.

I'd go with sentence case, but my comment here seeks consistency more than any particular casing.

(This is somewhat of a pre-existing condition, and not necessarily in scope to change, but does seem in scope to review as Git notes the line has changed.  This is not a blocker.)

**Comment by eweitz** [comment]
Trivial blocker: the [build failure](https://github.com/broadinstitute/single_cell_portal_core/actions/runs/5018231706/jobs/8997386517?pr=1781#step:7:4306) looks like a true positive.

```
### FAILURES and ERRORS ###
*** Yarn test failures ***

*** Rails test failures ***

  AnnDataIngestParametersTest#test_should_instantiate_and_validate_params [/home/app/webapp/test/models/ann_data_ingest_parameters_test.rb:65]:
Exiting with code: 1
Error: Process completed with exit code 1.
```

**Review by ehanna4 — APPROVED** [review]
Overall looks good - only thing is the test failure looks like a true failure (related to the change to use .gz ext)

File: app/models/ingest_job.rb [hunk 2]
```diff
@@ -559,6 +559,37 @@ def launch_subsample_jobs
...
+          Rails.logger.info "Launching subsampling ingest run for #{file_identifier} after #{action}"
+          submission = ApplicationController.papi_client.run_pipeline(
+            study_file:, user:, action: :ingest_subsample, params_object: subsample_params
+          )
+          IngestJob.new(
+            pipeline_name: submission.name, study:, study_file:, user:, action: :ingest_subsample,
+            params_object: subsample_params, reparse:, persist_on_fail:
+          ).poll_for_completion
+        end
+      end
```
[the remark this answers is not in the document]
Comment (bistline) [reply app/models/ingest_job.rb:592]: Agreed, though you're right in that it's out of scope.  But this method definitely is too long - I think a tech debt ticket for addressing issues like this is warranted.

File: app/models/ingest_job.rb (hunk 1 above)
Comment (bistline) [reply app/models/ingest_job.rb:673]: Yeah this is just good-old-fashioned laziness on my part...

File: app/lib/request_utils.rb [hunk 3]
```diff
@@ -115,6 +115,19 @@ def self.get_base_url
     "#{ApplicationController.default_url_options[:protocol]}://#{self.get_hostname}"
   end
 
+  # construct a GS URL or bucket path to a file extracted from an AnnData file
+  # fragment_type: cluster|metadata|matrix
+  # gs_url: true|false to prepend gs://{bucket_id} for absolute reference
+  # file_type_detail: either cluster obsm_key_name or processed|raw for matrix
```
[the remark this answers is not in the document]
Comment (bistline) [reply app/lib/request_utils.rb:121]: Ah shoot I missed this comment and the one about the Docker image.  Adding those changes now.

**Pull request merged**